File: Level 3/양과 늑대/solution.py
from collections import deque


def build_tree(info, edges):
    tree = [[] for _ in range(len(info))]

    for start, end in edges:
        tree[start].append(end)

    return tree


def solution(info, edges):
    tree = build_tree(info, edges)

    queue = deque([(0, 1, 0, set())])

    max_sheep = 0

    while queue:
        current, sheep_count, wolf_count, visited = queue.popleft()

        max_sheep = max(max_sheep, sheep_count)

        visited.update(tree[current])

        for next_node in visited:
            is_wolf = info[next_node]

            if not is_wolf:
                queue.append(
                    (next_node, sheep_count + 1, wolf_count, visited - {next_node})
                )
            elif is_wolf and sheep_count != wolf_count + 1:
                queue.append(
                    (next_node, sheep_count, wolf_count + 1, visited - {next_node})
                )

    return max_sheep

File: Level 3/양과 늑대/solution.js
function solution(info, edges) {
  const tree = buildTree(info, edges);

  let maxSheep = 0;

  const queue = [
    { current: 0, sheepCount: 1, wolfCount: 0, visited: new Set() },
  ];

  while (queue.length) {
    const { current, sheepCount, wolfCount, visited } = queue.shift();

    maxSheep = Math.max(maxSheep, sheepCount);

    tree[current].forEach((node) => visited.add(node));

    [...visited].forEach((nextNode) => {
      const isWolf = info[nextNode];

      const newVisited = new Set(
        [...visited].filter((node) => node !== nextNode)
      );
      if (!isWolf) {
        queue.push({
          current: nextNode,
          sheepCount: sheepCount + 1,
          wolfCount,
          visited: newVisited,
        });
      } else if (isWolf && sheepCount !== wolfCount + 1) {
        queue.push({
          current: nextNode,
          sheepCount,
          wolfCount: wolfCount + 1,
          visited: newVisited,
        });
      }
    });
  }

  return maxSheep;
}

function buildTree(info, edges) {
  const tree = Array.from({ length: info.length }, () => []);

  edges.forEach(([start, end]) => {
    tree[start].push(end);
  });

  return tree;
}
